web/src/ResourceGroupsContext.tsx

```diff
@@ -0,0 +1,81 @@
+import { createContext, PropsWithChildren, useContext } from "react"
+import { AnalyticsAction, AnalyticsType, incr } from "./analytics"
+import { usePersistentState } from "./LocalStorage"
+
+export type GroupState = { expanded: boolean }
+
+export type GroupsState = {
+  [key: string]: GroupState
+}
+
+type ResourceGroupsContext = {
+  groups: GroupsState
```

Comment:
i'm surprised you exposed both `groups` and `getGroups`? seems like you'd want people to use the accessor function.

Reply:
that's a good point! i thought it'd be useful for a context to still expose its whole state, but i don't have any strong opinions about it. (it is useful for testing.)